nixops/resources/__init__.py

```diff
@@ -3,12 +3,17 @@
 import re
 import nixops.util
 from threading import Event
-from typing import List, Optional, Dict, Any
+from typing import List, Optional, Dict, Any, Iterator
 from nixops.state import StateDict
 from nixops.diff import Diff, Handler
+from nixops.util import ImmutableMapping
 
 
-class ResourceDefinition(object):
+class ResourceOptions(ImmutableMapping[Any, Any]):
```

Comment:
I wonder if we can get some good juju from https://www.python.org/dev/peps/pep-0589/ here, somehow

Reply:
`TypedDict` certainly looks interesting but looking into it I'm not sure it's the correct choice (also not sure that it's _not_ the correct choice mind you).
`TypedDict` would give us type safety both at the _instantiation site_ and at the point of _usage_ using the familiar `__getitem__` interface.

If we want to add types to `ImmutableMapping` I think we may want to remove the `__getitem__` overload and only keep the `__getattr__` one and type it like any other object?
``` python
class MyMapping(ImmutableMapping):
    x: int
    y: str
```

An advantage of this approach could be composability:
``` python
class MyMappingBase(ImmutableMapping):
    x: int


class MyMapping(MyMappingBase):
    y: str


class SomeMapping(ImmutableMapping)
    aoeu: str
    submap: MyMapping


m = SomeMapping({
     "aoeu": "something",
     "submap": {  # submap is converted to a `MyMapping` in SomeMapping.__init__
       "x": 1,
       "y": "somethingelse",
     },
})


print(m.submap.x.lower())  # Mypy fails because `x` is an int
```
This approach would not give us type safety at _instantiation time_ (because we're passing a `dict`) but we could get some type safety out of this approach where the "mapping" is used.
My feeling is that we're very close to an immutable `dataclass` at this point (without the disadvantages that comes with subclassing a `dataclass`).

In this specific instance I don't think having type safety at instantiation gives us much since we're passing a dict from a Nix eval that we can't really know ahead of time what it contains.
That could be "solved" with some runtime reflection and throwing errors at runtime if we want to. Keep in mind that runtime checking of types does not work with generics so we'd lose that.